### PR TITLE
Tutorial for reading custom dataset 

### PR DESCRIPTION
**Pull request by pengyuan-zhou**

### CODE REVIEW

File: tutorial/Part X - Train on custom dataset.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Part X - Train on custom dataset
+
+A lot of effort in solving machine learning problem goes into data preparation. In this tutorial, we will show how to load and preprocess/augment data from a non trivial dataset using PyTorch tools.
+
+To give an example of this process, let's use the default demo-app and modify LocalMNISTDataDataSource.kt to load custom dataset.
+
+The overall logic of file reading is: 
+1. Read the files into butter [(link)](#Reader-init-and-restart)
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:8]: I think it's buffer and not butter ;p

File: tutorial/Part X - Train on custom dataset.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+2. Filter the data with demanded scope [(link)](#Custom-data-filter)
+3. Generate batches each consists of `batchSize` of samples [(link)](#Batching-and-sampling)
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:10]: consisting*

File: tutorial/Part X - Train on custom dataset.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import java.io.File
+```
+
+
+Define class and data readers.
+```kotlin
+class LocalMNISTDataDataSource constructor(
+private val resources: Resources
+) {
+private var trainDataReader = returnDataReader()
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:34]: Adding indentation will help for clarity

File: tutorial/Part X - Train on custom dataset.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+Define the paths to train file and label file.
+If the training data and labels are in the same file, then define the paths with same value.
+```kotlin
+private var trainFile = File("/path/to/train.csv")
+private var labelFile = File("/path/to/label.csv")
+```
+
+
+If the files locate in external storage, add following lines to AndroidManifest.xml
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:47]: This and the next column should come under a single header so that the user understands this step and the step above are replaceable

File: tutorial/Part X - Train on custom dataset.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+And read the files as follows
+```kotlin
+private var trainFile = File(Environment.getExternalStorageDirectory()+"/path/to/train.csv")
+private var labelFile = File(Environment.getExternalStorageDirectory()+"/path/to/label.csv")
+```
+
+
+Define the dimensions and starting indices of training data and labels, and the delimiter of the files.
+```kotlin
+private var trainDim = TD
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:63]: May be tell what is TD

File: tutorial/Part X - Train on custom dataset.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+private var labelDim = LD
+private var trainStart = TS
+private var labelStart = LS
+private var delim = ","
+```
+
+
+#### Batching and sampling.
+```kotlin
+fun loadDataBatch(batchSize: Int): Pair<Batch, Batch> {
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:73]: Indentation is needed here

**Comment by iamtrask** [comment]
I am not the primary gatekeeper for this repository, so let's wait for other approvals before merging, but it's a very nice and valuable contribution from my perspective :)

File: tutorial/Part X - Train on custom dataset.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+longArrayOf(trainInput.size.toLong(), trainDim)
+)
+val trainingLabel = Batch(
+labels.flatten().toFloatArray(),
+longArrayOf(labels.size.toLong(), labelDim)
+)
+return Pair(trainingData, trainingLabel)
+}
+
+private fun readSample(
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:90]: leave a comment here saying readSample reads a single line from data and label files and appends them into respective lists. These lists are then merged later to form a batch of data

File: tutorial/Part X - Train on custom dataset.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+sample.second.map { it.trim().toFloat() }
+)
+}
+```
+
+
+#### Custom data filter.
+```kotlin
+private fun readLine(): Pair<List<String>, List<String>> {
+var x = trainDataReader.readLine()?.split(delim)
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:109]: readline or go back to the beginning of the file if we have exhausted the dataset. This allows us to iterate infinitely over the same data

File: tutorial/Part X - Train on custom dataset.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
+# Part X - Train on custom dataset
+
+A lot of effort in solving machine learning problem goes into data preparation. In this tutorial, we will show how to load and preprocess/augment data from a non trivial dataset using PyTorch tools.
+
+To give an example of this process, let's use the default demo-app and modify LocalMNISTDataDataSource.kt to load custom dataset.
```
Comment (vkkhare) [inline tutorial/Part X - Train on custom dataset.md:5]: Also attach a link to LocalMNISTDataDataSource.kt

File: tutorial/Part X - Train on custom dataset.md (hunk 8 above)
Comment (pengyuan-zhou) [reply tutorial/Part X - Train on custom dataset.md:109]: hi, do you mean to add this line as a comment there?

Comment (vkkhare) [reply tutorial/Part X - Train on custom dataset.md]: Yeah inline comment will work

**Comment by vkkhare** [comment]
Lgtm

**Comment by vkkhare** [comment]
@all-contributors add @pengyuan-zhou for writing tutorials
